content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,1422 @@
+          // Reference: https://xrpl.org/subscribe.html
+          await client.request({
+            "command": "subscribe",
+            "streams": ["ledger"]
+          })
+
+          // Dispatch 'update-ledger-data' event
+          client.on("ledgerClosed", async (ledger) => {
+            appWindow.webContents.send('update-ledger-data', ledger)
+          })
+        }
+
+app.whenReady().then(main)
+```
+
+Here, we have reduced the `app.whenReady` logic to an oneliner and put the necessary functionality into a separate `main()` function. The most relevant piece of code here is the swapping of a single call to the ledger for a subscription: Our client is now connecting to the XRPL via [WebSockets](https://en.wikipedia.org/wiki/WebSocket). This establishes a permanent bidirectional connection to the XRPL, which allows us to subscribe to events that the server sends out. This saves resources on the server, which now only sends out data we explicitly asked for when a change happens, as well as the client which does not have to sort through incoming data for relevant changes. This also reduces the complexity of the application and saves us a couple of lines of code.
+
+2. Then, update `preload.js` by renaming the `onUpdateLedgerIndex` to `onUpdateLedgerData` and the `update-ledger-index` event to `update-ledger-data`:
+
+```javascript
+const { contextBridge, ipcRenderer } = require('electron');
+
+contextBridge.exposeInMainWorld('electronAPI', {
+  onUpdateLedgerData: (callback) => {
+    ipcRenderer.on('update-ledger-data', callback)
+  }
+})
+```
+
+This renaming might seem a bit nit-picky, but now we actually pass on an object of values instead of a single integer.
+
+3. Next, modify the `view/template.html` file by adding placeholders to the `<body>` section:
+
+```html
+<body>
+    <!-- Step 2 code additions - start -->
+    <h3>Build a XRPL Wallet - Part 2/8</h3>
+    Latest validated ledger <br />
+    Ledger Index: <strong id="ledger-index"></strong><br />
+    Ledger Hash: <strong id="ledger-hash"></strong><br />
+    Close Time: <strong id="ledger-close-time"></strong><br />
+    <!-- Step 2 code additions - end -->
+</body>
+```
+
+4. Modify the `view/render.js` file to handle the new placeholders. Note that the renamed preloader function is now reflected in `window.electronAPI.onUpdateLedgerData`:
+
+```javascript
+const ledgerIndexEl = document.getElementById('ledger-index')
+const ledgerHashEl = document.getElementById('ledger-hash')
+const ledgerCloseTimeEl = document.getElementById('ledger-close-time')
+
+window.electronAPI.onUpdateLedgerData((_event, value) => {
+    ledgerIndexEl.innerText = value.ledger_index
+    ledgerHashEl.innerText = value.ledger_hash
+    ledgerCloseTimeEl.innerText = value.ledger_time
+})
+```
+
+This should make our application listen to regular updates of the ledger and display them in the frontend.
+
+Now run the application with the following command:
+
+```console
+npm run start
+```
+
```

Comment:
At this point "Latest validated ledger" seems like it should have information, but it's just there as a title. It should probably be an <h3> or something to differentiate it from the rest of the information.
![Uploading image.png…]()